test/common/PrepDataFuncs.cpp

```diff
@@ -0,0 +1,342 @@
+                                         collArgs.root,
+                                         false);
+  }
+
+  ErrCode DefaultPrepData_Reduce(CollectiveArgs &collArgs, bool const isAllReduce)
+  {
+    if (CheckAllocation(collArgs) != TEST_SUCCESS) return TEST_FAIL;
+    if (collArgs.numInputElements != collArgs.numOutputElements)
+    {
+      ERROR("Number of input elements must match number of output elements for Reduce\n");
+      return TEST_FAIL;
+    }
+
+    size_t const numBytes = collArgs.numInputElements * DataTypeToBytes(collArgs.dataType);
+
+    // Clear output for all ranks (done before filling input in case of in-place)
+    CHECK_CALL(collArgs.outputGpu.ClearGpuMem(numBytes));
+
+    // Clear expected buffer for holding reduction
+    PtrUnion result;
+    CHECK_CALL(result.Attach(collArgs.expected));
+    CHECK_CALL(result.ClearCpuMem(numBytes));
+
+    // If average or custom reduction operator is used, perform a summation instead
+    ncclRedOp_t const tempOp = (collArgs.redOp >= ncclAvg ? ncclSum : collArgs.redOp);
+
+    // Loop over each rank and generate their input into a temp buffer, then reduce
+    PtrUnion scalarsPerRank;
+    scalarsPerRank.Attach(collArgs.scalarTransport.ptr);
+
+    PtrUnion tempInputCpu;
+    CHECK_CALL(tempInputCpu.Attach(collArgs.outputCpu));
+    for (int rank = 0; rank < collArgs.totalRanks; ++rank)
+    {
+      // Generate temporary input for this rank
+      CHECK_CALL(tempInputCpu.FillPattern(collArgs.dataType, collArgs.numInputElements, rank, false));
+
+      // Copy the pre-scaled input into GPU memory for the correct rank
+      if (rank == collArgs.globalRank)
+      {
+        CHECK_HIP(hipMemcpy(collArgs.inputGpu.ptr, tempInputCpu.ptr, numBytes, hipMemcpyHostToDevice));
+      }
+
+      // Scale the temporary input by local scalar for this rank
+      // (Used by custom reduction ops)
+      if (collArgs.scalarMode >= 0)
+      {
+        CHECK_CALL(tempInputCpu.Scale(collArgs.dataType, collArgs.numInputElements,
+                                      scalarsPerRank, rank));
+      }
+
+      // Any rank that requires output reduces the scaled-inputs
+      if (isAllReduce || collArgs.root == collArgs.globalRank)
+      {
+        if (rank == 0)
+        {
+          memcpy(result.ptr, tempInputCpu.ptr, numBytes);
+        }
+        else
+        {
+          CHECK_CALL(result.Reduce(collArgs.dataType, collArgs.numInputElements,
+                                   tempInputCpu, tempOp));
+        }
+      }
+    }
+
+    // Perform averaging if necessary
+    if (collArgs.redOp == ncclAvg && (isAllReduce || collArgs.root == collArgs.globalRank))
+    {
+      CHECK_CALL(result.DivideByInt(collArgs.dataType, collArgs.numInputElements, collArgs.totalRanks));
+    }
+    return TEST_SUCCESS;
+  }
+
+  ErrCode DefaultPrepData_Gather(CollectiveArgs &collArgs, bool const isAllGather)
+  {
+    if (CheckAllocation(collArgs) != TEST_SUCCESS) return TEST_FAIL;
```

Comment:
`CHECK_CALL(CheckAllocation(collArgs)) ` same as above (next time it happens again too)